docs/spec.md

```diff
@@ -2813,8 +2814,9 @@ More formally, `result[result_index] = operand[operand_index]` where:
     `index_vector_dim` < `rank(start_indices)`.
   * `[start_indices[batch_index]]` otherwise.
 * For `d_operand` in `axes(operand)`,
-  * `full_start_index[d_operand]` = `start_index[d_start]` if
-    `d_operand = start_index_map[d_start]`.
+  * `full_start_index[d_operand]` =
+    `clamp(start_index[d_start], 0, dim(operand, d_operand) - slice_sizes[d_operand])`
+    if `d_operand = start_index_map[d_start]`.
```

Comment:
Why do we need to clamp specifically here instead of e.g. at the very last step?

Reply:
We still want to keep the information about how much to offset for `fullOffsetIndex`. By clamping after the addition, we lose this information. For example, if `start_index` had -100, and offset 2, we would get `clamp(min=0, -100+2, max) = min` whereas we want (min+2) as the index.

Reply:
"whereas we want (min+2) as the index". To dig a bit deeper, can you share a reference to documentation or code that depends on / implements this behavior?

Reply:
Perhaps this code snippet is what you're looking for?https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/hlo/evaluator/hlo_evaluator.cc#L2635-L2644

Operation Semantics do not explicitly state clamping, but the comments in the code state the clamping before addition.

Reply:
A link to hlo_evaluator is useful, but since this semantics is not part of https://www.tensorflow.org/xla/operation_semantics#gather, we'd need a bit more. Otherwise, it's hard to say whether the clamping behavior is perhaps implementation-defined behavior. 

Can we have some additional paper trail, e.g. links to some frameworks making use of this semantics (e.g. JAX) or testimonials from representatives of these frameworks?

Reply:
Here is the link to JAX docs on gather: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.gather.html
Note for JAX, the mode can be one of three values `{CLIP, FILL_OR_DROP, PROMISE_IN_BOUNDS}`. 
https://jax.readthedocs.io/en/latest/jax.lax.html#jax.lax.GatherScatterMode

PROMISE_IN_BOUNDS:
The user promises that indices are in bounds. No additional checking will be performed. In practice, with the current XLA implementation this means that, out-of-bounds gathers will be clamped but out-of-bounds scatters will be discarded. Gradients will not be correct if indices are out-of-bounds.

Reply:
I don't think that "indices are in bounds" and "out-of-bounds gathers will be clamped" indicate specifically that clamping should happen at the `full_start_index` level and not at the `operand_index` level?

Reply:
Here is another [reference](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/elemental_ir_emitter.cc#L2587-L2602) where this behavior is similarly replicated.